### PR TITLE
Pass Semi sync information from durability policy and use it to log discrepancies

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

In release 13, we will introduce a new flag `durability` in vtctld and vtctl. That durability policy should be the one that reflects the semi-sync settings and we will eventually deprecate `enable_semi_sync` flag on the vttablets. But all of this cannot be accomplished in a single release because of upgrade downgrade considerations. So the plan is as follows - 
- Introduce the new flag `durability`
- Pass the information all through the components and just document the discrepancies in  configuration.
- Keep using `enable_semi_sync` flag in v13, so the new flag is only used for logging but it is expected that vtctl, vtctld binaries of v13 will have the flag set.
- In release 14, we would have the semi-sync information being passed everywhere, so we can deprecate `enable_semi_sync` and use the information coming from the durability policies.

This PR accomplishes this change of passing the information down and logging the differences for release 13. 

## Related Issue(s)

- #8975 

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [ ] Documentation was added or is not required - Pending

### CODE REVIEW

**Comment by GuptaManan100** [comment]
Apart from the checks in this PR, I also made a change wherein we start using the durability policy information in `fixSemiSync` in #9552. That PR is 2 commits ahead of this PR - 
1. https://github.com/vitessio/vitess/pull/9552/commits/1380e76004e394a264f637b647ac8c6d45228ba4 wherein we change fixSemiSyncReplication
2. https://github.com/vitessio/vitess/pull/9552/commits/39e6f09c973d943cfcb97d0a9f2d5db3a71f34e5 which fixes some of the unit tests to set the correct durability policies and skips the backup tests.

On that PR, it can be observed that the only failing tests are from `cluster 11, 16, 19, 20, 21, 26, docker_test_cluster_10` which all run `Backup` RPC and `Upgrade Downgrade test for reparenting with old vtctl`. Their are 2 failing tests, one `TestSemiSyncSetupCorrectly` and `TestChangeTypeSemiSync` which both test semi-sync settings. We expect these tests to fail since old vtclt does not have durability policy information and the new vttablets in that branch only fixSemiSync using the parameter passed down. The failure in tests concerned with `Backup` RPC is also expected, since we will no longer be fixing semiSync after a `Backup` operation succeeds. We will rely on vtorc to fix it. The main reason for this -
  1. After the durability policy changes take effect, a tablet might be sending semi sync acks to one primary but may be configured to not send semi sync acks to some other primary, and since `Backup` is a long operation during which the primary might have changed, we will have no idea whether we should be setting our semi-sync or not. So we will rely on vtorc to fix it when it becomes a compulsory component of Vitess. An alternate is to run StartReplication RPC after running Backup RPC, and that will figure out the semi sync settings.

File: go/vt/orchestrator/inst/tablet_dao.go [hunk 1]
```diff
@@ -80,15 +80,15 @@ func SwitchPrimary(newPrimaryKey, oldPrimaryKey InstanceKey) error {
 		log.Errore(err)
 		return nil
 	}
-	if _, err := ChangeTabletType(oldPrimaryKey, topodatapb.TabletType_REPLICA); err != nil {
+	if _, err := ChangeTabletType(oldPrimaryKey, topodatapb.TabletType_REPLICA, ReplicaSemiSync(newPrimaryKey, oldPrimaryKey)); err != nil {
```
Comment (shlomi-noach) [inline go/vt/orchestrator/inst/tablet_dao.go:83]: The function name `ReplicaSemiSync` is confusing. I'd be more comfortable if it were called `IsReplicaSemiSync`

File: go/vt/vttablet/tabletmanager/rpc_actions.go [hunk 2]
```diff
@@ -43,6 +43,17 @@ const (
...
 )
 
+// SemiSyncAction is used to tell fixSemiSync whether to change the semi-sync
+// settings or not.
+type SemiSyncAction int
+
+// Allowed values for SemiSyncAction
+const (
+	SemiSyncActionNone = SemiSyncAction(iota)
+	SemiSyncActionTrue
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletmanager/rpc_actions.go:53]: Perhaps name as `SemiSyncActionSet`?

File: go/vt/vttablet/tabletmanager/rpc_actions.go [hunk 3]
```diff
@@ -43,6 +43,17 @@ const (
...
+	SemiSyncActionFalse
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletmanager/rpc_actions.go:54]: Perhaps name as `SemiSyncActionUnset`?

**Review by shlomi-noach — COMMENTED** [review]
I don't have much quality review content for now, just a few naming suggestions.

File: go/vt/vttablet/tabletmanager/rpc_replication.go [hunk 4]
```diff
@@ -475,18 +475,20 @@ func (tm *TabletManager) demotePrimary(ctx context.Context, revertPartialFailure
...
-	if err := tm.fixSemiSync(topodatapb.TabletType_REPLICA); err != nil {
-		return nil, err
-	}
-	defer func() {
-		if finalErr != nil && revertPartialFailure && wasPrimary {
-			// enable primary-side semi-sync again
-			if err := tm.fixSemiSync(topodatapb.TabletType_PRIMARY); err != nil {
-				log.Warningf("fixSemiSync(PRIMARY) failed during revert: %v", err)
-			}
+	if tm.isPrimarySideSemiSyncEnabled() {
```
Comment (deepthi) [inline go/vt/vttablet/tabletmanager/rpc_replication.go:478]: Can you add a comment here as to why we should check this first?

File: go/vt/worker/multi_split_diff.go [hunk 5]
```diff
@@ -407,7 +407,7 @@ func (msdw *MultiSplitDiffWorker) stopReplicationOnSourceTabletAt(ctx context.Co
 		msdw.wr.Logger().Infof("stopping Replication %v at a minimum of %v", msdw.sourceAlias, vreplicationPos)
 
 		shortCtx, cancel = context.WithTimeout(ctx, *remoteActionsTimeout)
-		msdw.wr.TabletManagerClient().StartReplication(shortCtx, sourceTablet.Tablet)
+		msdw.wr.StartReplication(shortCtx, sourceTablet.Tablet)
```
Comment (deepthi) [inline go/vt/worker/multi_split_diff.go:410]: This is brilliant. MultiSplitDiffWorker is created using a vtctld command which means we have the durability policy and wrangler can enforce it. 💯

File: web/vtadmin/src/proto/vtadmin.d.ts [hunk 6]
```diff
@@ -43087,6 +43135,12 @@ export namespace binlogdata {
 
         /** VEvent last_p_k_event */
         last_p_k_event?: (binlogdata.ILastPKEvent|null);
+
+        /** VEvent keyspace */
+        keyspace?: (string|null);
+
+        /** VEvent shard */
+        shard?: (string|null);
```
Comment (deepthi) [inline web/vtadmin/src/proto/vtadmin.d.ts:43143]: @doeg @ajm188 these fields were added to the proto definition some time ago, but the vtadmin files don't seem to have been regenerated at that time.

File: web/vtadmin/src/proto/vtadmin.js [hunk 7]
```diff
@@ -30842,6 +30866,14 @@ $root.tabletmanagerdata = (function() {
...
 
+        /**
+         * ApplySchemaRequest sql_mode.
+         * @member {string} sql_mode
+         * @memberof tabletmanagerdata.ApplySchemaRequest
+         * @instance
+         */
+        ApplySchemaRequest.prototype.sql_mode = "";
+
         /**
```
Comment (deepthi) [inline web/vtadmin/src/proto/vtadmin.js:30877]: This seems to be an artifact of a different PR as well.

File: web/vtadmin/src/proto/vtadmin.d.ts (hunk 6 above)
Comment (deepthi) [reply web/vtadmin/src/proto/vtadmin.d.ts:43143]: Can one of you post reminders in #maintainers and #developers so that people don't miss the step of generating vtadmin files?

File: web/vtadmin/src/proto/vtadmin.js [hunk 8]
```diff
@@ -35887,6 +35932,7 @@ $root.tabletmanagerdata = (function() {
          * Properties of a StartReplicationRequest.
          * @memberof tabletmanagerdata
          * @interface IStartReplicationRequest
+         * @property {boolean|null} [semiSync] StartReplicationRequest semiSync
```
Comment (deepthi) [inline web/vtadmin/src/proto/vtadmin.js:35935]: @notfelineit do these changes affect the "destructive" tablet actions UI?

**Review by deepthi — APPROVED** [review]
Nice work.
Please address all feedback and then this can be merged.

File: go/vt/orchestrator/inst/tablet_dao.go (hunk 1 above)
Comment (GuptaManan100) [reply go/vt/orchestrator/inst/tablet_dao.go:83]: Changed

File: go/vt/vttablet/tabletmanager/rpc_actions.go (hunk 2 above)
Comment (GuptaManan100) [reply go/vt/vttablet/tabletmanager/rpc_actions.go:53]: Renamed

File: go/vt/vttablet/tabletmanager/rpc_actions.go (hunk 3 above)
Comment (GuptaManan100) [reply go/vt/vttablet/tabletmanager/rpc_actions.go:54]: Renamed

File: go/cmd/vtworker/vtworker.go [hunk 9]
```diff
@@ -37,13 +37,15 @@ import (
...
 	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/vtctl/reparentutil"
 	"vitess.io/vitess/go/vt/worker"
 )
 
 var (
 	cell                   = flag.String("cell", "", "cell to pick servers from")
 	commandDisplayInterval = flag.Duration("command_display_interval", time.Second, "Interval between each status update when vtworker is executing a single command from the command line")
 	username               = flag.String("username", "", "If set, value is set as immediate caller id in the request and used by vttablet for TableACL check")
+	durability             = flag.String("durability", "none", "type of durability to enforce. Default is none. Other values are dictated by registered plugins")
```
[the remark this answers is not in the document]
Comment (GuptaManan100) [reply go/cmd/vtworker/vtworker.go:48]: Fixed, it is good we did this now, otherwise we would have had to go through 2 more releases to fix it later 😆

File: go/vt/worker/multi_split_diff.go (hunk 5 above)
Comment (GuptaManan100) [reply go/vt/worker/multi_split_diff.go:410]: It is actually created from vtworker, but we have added durability policy to that too! So everything works now!

File: go/vt/vttablet/tabletmanager/rpc_replication.go (hunk 4 above)
Comment (GuptaManan100) [reply go/vt/vttablet/tabletmanager/rpc_replication.go:480]: Added

File: web/vtadmin/src/proto/vtadmin.d.ts (hunk 6 above)
Comment (doeg) [reply web/vtadmin/src/proto/vtadmin.d.ts:43143]: I've posted another reminder to both channels. (This was also mentioned back at the beginning of December.) I expect this will (understandably) be an issue until we can enforce this in CI. The ticket for that is https://github.com/vitessio/vitess/issues/9571.

File: web/vtadmin/src/proto/vtadmin.js (hunk 8 above)
Comment (notfelineit) [reply web/vtadmin/src/proto/vtadmin.js:35935]: I don't think so, not as it exists - should we add a semiSync/durability option in the UI?

Comment (notfelineit) [reply web/vtadmin/src/proto/vtadmin.js:35935]: I don't think so, not as it exists - should we add a semiSync/durability option in the UI?

Comment (notfelineit) [reply web/vtadmin/src/proto/vtadmin.js:35935]: Screen Shot 2022-01-27 at 9 45 23 AM
Yes! It follows vtctld2's ui in that there is a singular button

Comment (deepthi) [reply web/vtadmin/src/proto/vtadmin.js:35935]: I suspect that this directly calls the TabletManager grpc API instead of going through vtctld API.
My suggestion is to add a new vtctld API. Then semi_sync won't need to be provided by the user, it can be computed by VtctldServer.

Comment (notfelineit) [reply web/vtadmin/src/proto/vtadmin.js:35935]: I'll check today and see which it is using - and switch to vtctld if necesary.

Comment (notfelineit) [reply web/vtadmin/src/proto/vtadmin.js:35935]: @deepthi vtadmin uses this codepath: https://cs.github.com/vitessio/vitess/blob/f906693c6441f4b41d8f842ffb9dba85258b5158/go/vt/vtctl/grpcvtctldserver/server.go?q=StartReplication#L2439 it does call the tablet manager API, but I think this is the vtctld API mentioned?
